src/Korobi/WebBundle/Util/GitInfo.php

```diff
@@ -13,45 +13,130 @@ class GitInfo {
+        /** @var $firstItem Callable */
+        $this->$firstItem($gitDir, $process);
+
+        // Handle capifony
+        $this->postCheckForCapifonyFiles($gitDir, $environment);
+    }
+
+    /*
+     * Begin chain-of-responsibility function definitions
+     * --------------------------------------------------
+     */
+
+    private function checkIsGitRepo($gitDir, $stack) {
+        if (file_exists($gitDir)) {
+            /** @var Callable $next */
+            $next = array_shift($stack);
+            $this->$next($gitDir, $stack);
+        } else {
+            $this->setBranchAndHash("unknown", "unknown");
+        }
+    }
+
+    private function checkForGitHeadFile($gitDir, $stack) {
+        if (file_exists($gitDir . "HEAD")) {
+            $headContents = fgets(fopen($gitDir . "HEAD", 'r')); // resource acquisition is initialization
+            $this->handleGitHeadFile($gitDir, $headContents, $stack);
+        } else {
+            $this->setBranchAndHash("unknown", "unknown");
+        }
+    }
+
+    private function handleGitHeadFile($gitDir, $headContents, $stack) {
+        if (strpos($headContents, "ref: ") === -1) {
+            $this->setBranchAndHash($headContents, $headContents);
+        } else {
+            $referencedFile = trim(substr($headContents, 5));
+            $this->branch = str_replace("refs/heads/", "", $referencedFile);
+            $this->tempRef = $referencedFile;
+            if (!file_exists($gitDir . $referencedFile)) {
+                /** @var Callable $next */
+                $next = array_shift($stack);
+                $this->$next($gitDir, $stack);
             } else {
-                $ref = trim(file_get_contents($rootDir . '/../.git/HEAD'));
-                $this->branch = str_replace('refs/heads/', '', explode(' ', $ref)[1]);
+                $this->hash = file_get_contents($gitDir . $referencedFile);
             }
-            $this->hash = trim(file_get_contents($rootDir . '/../REVISION'));
-            return;
         }
+    }
 
-        // Accurate information retrieval for local environments
-        $branch = trim(`git rev-parse --abbrev-ref HEAD 2>&1`);
-        if (StringUtil::startsWith($branch, 'fatal: Not a git repository')) {
-            $this->branch = 'unknown';
-            $this->hash = 'unknown';
-            return;
+    private function checkForPackedRefs($gitDir, $stack) {
+        $refToLocate = $this->tempRef;
+        if (file_exists($gitDir . "packed-refs")) {
+            $packedRefData = file_get_contents($gitDir . "packed-refs");
+            $loc = strpos($packedRefData, $refToLocate) - 1;
+            $commitHash = "";
+            if ($loc !== false) {
+                while ($packedRefData[$loc] !== "\n" && $loc > 0) {
+                    $commitHash .= $packedRefData[$loc];
+                    $loc--;
+                }
+                $this->hash = strrev($commitHash);
+                return;
+            }
         }
```

Comment:
This looks overly complex to me, why not `file($gitDir . "packed-refs")` and `foreach` combined with `strpos`?